provider/simple/reprovide.go

```diff
@@ -0,0 +1,225 @@
+package simple
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	backoff "github.com/cenkalti/backoff"
+	cid "github.com/ipfs/go-cid"
+	cidutil "github.com/ipfs/go-cidutil"
+	blocks "github.com/ipfs/go-ipfs-blockstore"
+	pin "github.com/ipfs/go-ipfs/pin"
+	ipld "github.com/ipfs/go-ipld-format"
+	logging "github.com/ipfs/go-log"
+	merkledag "github.com/ipfs/go-merkledag"
+	verifcid "github.com/ipfs/go-verifcid"
+	routing "github.com/libp2p/go-libp2p-routing"
+)
+
+var logR = logging.Logger("reprovider.simple")
+
+//KeyChanFunc is function streaming CIDs to pass to content routing
+type KeyChanFunc func(context.Context) (<-chan cid.Cid, error)
+type doneFunc func(error)
+
+// Reprovider reannounces blocks to the network
+type Reprovider struct {
+	ctx     context.Context
+	trigger chan doneFunc
+
+	// The routing system to provide values through
+	rsys routing.ContentRouting
+
+	keyProvider KeyChanFunc
+
+	tick time.Duration
+}
+
+// NewReprovider creates new Reprovider instance.
+func NewReprovider(ctx context.Context, reprovideIniterval time.Duration, rsys routing.ContentRouting, keyProvider KeyChanFunc) *Reprovider {
+	return &Reprovider{
+		ctx:     ctx,
+		trigger: make(chan doneFunc),
+
+		rsys:        rsys,
+		keyProvider: keyProvider,
+		tick:        reprovideIniterval,
+	}
+}
+
+// Close the reprovider
+func (rp *Reprovider) Close() error {
+	return nil
```

Comment:
This probably should close `rp.ctx`

Reply:
I can add this if we want, but my goal was to change the existing system as little as possible. Not closing `rp.ctx` means it continues behaving the way it has, which I'm OK with since my goal is to replace the whole system and don't want to unintentionally cause a problem that diverts my energy. What do you think?

Reply:
It's probably fine